src/Umbraco.New.Cms.Core/Factories/UpgradeSettingsFactory.cs

```diff
@@ -0,0 +1,34 @@
+﻿using Umbraco.Cms.Core.Configuration;
+using Umbraco.Cms.Core.Semver;
+using Umbraco.Cms.Core.Services;
+using Umbraco.New.Cms.Core.Models.Installer;
+
+namespace Umbraco.New.Cms.Core.Factories;
+
+public class UpgradeSettingsFactory : IUpgradeSettingsFactory
+{
+    private readonly IRuntimeState _runtimeState;
+    private readonly IUmbracoVersion _umbracoVersion;
+
+    public UpgradeSettingsFactory(
+        IRuntimeState runtimeState,
+        IUmbracoVersion umbracoVersion)
+    {
+        _runtimeState = runtimeState;
+        _umbracoVersion = umbracoVersion;
+    }
+
+
+    public UpgradeSettingsModel GetUpgradeSettings()
+    {
+        var model = new UpgradeSettingsModel
+        {
+            CurrentState = _runtimeState.CurrentMigrationState ?? string.Empty,
+            NewState = _runtimeState.FinalMigrationState ?? string.Empty,
+            NewVersion = _umbracoVersion.SemanticVersion,
+            OldVersion = new SemVersion(_umbracoVersion.SemanticVersion.Major),
```

Comment:
`OldVersion` is likely not correct here. Consider re-introducing the TODO from the original source of this (https://github.com/umbraco/Umbraco-CMS/blob/fd2c9de939408e760003aa88623000bf93e3b5fb/src/Umbraco.Core/Install/InstallSteps/UpgradeStep.cs#L48)